### PR TITLE
[X86][CodeGen] Transform NDD SUB to CMP if dest reg is dead

### PR DESCRIPTION
**Pull request by KanRobert**

### CODE REVIEW

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 1]
```diff
@@ -5130,39 +5162,51 @@ bool X86InstrInfo::optimizeCompareInstr(MachineInstr &CmpInstr, Register SrcReg,
     default:
       llvm_unreachable("Unreachable!");
     case X86::SUB64rm:
+    case X86::SUB64rm_ND:
       NewOpcode = X86::CMP64rm;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:5166]: Will it be a problem if `SUB64rm_ND` is assigned with EGPR reigsters?

**Review by e-kud — APPROVED** [review]
LGTM. Thanks!

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:5166]: Do we need to consider `SUB64rm_EVEX`?

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:5166]: No, it's a redundant encoding and never generated by compiler.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by KanRobert** [comment]
Thanks @e-kud @phoebewang !

**Pull request merged**